### PR TITLE
Eliminate TensorIterator::add_output with explicit dtype.

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39803 Split TensorIteratorConfig out of TensorIterator
* **#39800 Eliminate TensorIterator::add_output with explicit dtype.**

I'm working on a refactor where I want to represent the inputs
and outputs to TensorIterator as just plain Tensors, which means
I need to kill this add_output with explicit dtype.  This exists
solely to set what the output dtype should be.  We have a pretty
similar API for doing this for shapes (declare_static_shape) so
I just copied this API for dtypes instead.

Although the new version is more code, I think the intent is more
explicit.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21981740](https://our.internmc.facebook.com/intern/diff/D21981740)

### CODE REVIEW

**Review by mruberry — APPROVED** [review]
Good simplification.

**Review by robieta — APPROVED** [review]
LGTM.

**Review by VitalyFedyunin — APPROVED** [review]

**Comment by ezyang** [comment]
CI failure is legit, somehow.

File: aten/src/ATen/native/TensorIterator.h [hunk 1]
```diff
@@ -372,6 +367,12 @@ struct CAFFE2_API TensorIterator {
     config_resize_outputs_ = false;
   }
 
+  // Bypass output dtype computation and fix the dtype as specified here.
+  void declare_static_dtype(ScalarType dtype) {
+    TORCH_CHECK(!check_all_same_dtype(), "check_all_same_dtype(false) must be called before declare_static_dtype(...)");
```
[the remark this answers is not in the document]
Comment (ezyang) [reply aten/src/ATen/native/TensorIterator.h:372]: I'm just going to slink into the shadows mumbling "well, if you treat TensorIterator as the public API..."

**Pull request closed without merging**